File: apps/api/src/application/support/auth-service.ts
import * as jose from 'jose';
import { adminUsers } from '../../db/schema.js';

export type AdminPayload = {
	sub: string;
	email: string;
	name: string;
};

const getSecret = () => {
	const s = process.env.JWT_SECRET;
	if (!s || s.length < 16) {
		throw new Error('JWT_SECRET must be set (min 16 chars)');
	}
	return new TextEncoder().encode(s);
};

export class AuthService {
	async signAdminToken(user: typeof adminUsers.$inferSelect): Promise<string> {
		const secret = getSecret();
		return new jose.SignJWT({
			email: user.email,
			name: user.name
		})
			.setProtectedHeader({ alg: 'HS256' })
			.setSubject(String(user.id))
			.setIssuedAt()
			.setExpirationTime('7d')
			.sign(secret);
	}

	async verifyAdminToken(token: string | undefined): Promise<AdminPayload | null> {
		if (!token) return null;
		try {
			const secret = getSecret();
			const { payload } = await jose.jwtVerify(token, secret);
			const sub = payload.sub;
			const email = payload.email;
			const name = payload.name;
			if (typeof sub !== 'string' || typeof email !== 'string' || typeof name !== 'string') {
				return null;
			}
			return { sub, email, name };
		} catch {
			return null;
		}
	}
}

File: apps/python-api/src/yourmedportal_api/application/support/auth_service.py
from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Any

import jwt

from yourmedportal_api.config import get_settings


@dataclass
class AdminPayload:
    sub: str
    email: str
    name: str


class AuthService:
    def sign_admin_token(self, user_id: int, email: str, name: str) -> str:
        s = get_settings()
        if not s.JWT_SECRET or len(s.JWT_SECRET) < 16:
            raise RuntimeError("JWT_SECRET must be set (min 16 chars)")
        now = int(time.time())
        payload: dict[str, Any] = {
            "sub": str(user_id),
            "email": email,
            "name": name,
            "iat": now,
            "exp": now + 7 * 24 * 3600,
        }
        token = jwt.encode(payload, s.JWT_SECRET, algorithm="HS256")
        if isinstance(token, bytes):
            return token.decode("ascii")
        return str(token)

    def verify_admin_token(self, token: str | None) -> AdminPayload | None:
        if not token:
            return None
        s = get_settings()
        if not s.JWT_SECRET:
            return None
        try:
            decoded = jwt.decode(token, s.JWT_SECRET, algorithms=["HS256"])
            sub = decoded.get("sub")
            email = decoded.get("email")
            name = decoded.get("name")
            if not isinstance(sub, str) or not isinstance(email, str) or not isinstance(name, str):
                return None
            return AdminPayload(sub=sub, email=email, name=name)
        except Exception:
            return None
